Improve visualize modal test coverage

graceguo-supercat opened the pull request:
fixes https://github.com/airbnb/superset/pull/2849

Commit:
redux visualize modal

Commit:
redux visualize modal

- apply redux
- add unit test for connect container component
- fix lint error

Commit:
add unit tests for VisuaizeModal

Commit:
add unit tests for VisualizeModal component

Commit:
Merge remote-tracking branch 'supercat/gg-addTestCoverage-VisualizeModal' into gg-addTestCoverage-VisualizeModal

# Conflicts:
#	superset/assets/javascripts/SqlLab/actions.js

graceguo-supercat commented:
Add more unit tests for VisualModal component to increase test coverage.

Commit:
remove changes on css file (merge by mistake)

Commit:
remove changes on css file (merge by mistake)

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
-      chartType: mockChartTypeBarChart,
-      columns: mockColumns,
-      datasourceName: 'mockDatasourceName',
+    it('should have valid query prop', () => {
```
this test doesn't look like it's asserting that there is a query prop...

graceguo-supercat replied:
setStateFromProps required valid props as parameter, otherwise it should not change state.
So I add a test that pass invalid paramters, and expects no state change.

The old test description is mis-leading. I will give it a meaningful description.

Commit:
update unit test description

setStateFromProps required valid props as parameter, otherwise it should not change state.

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
+      });
+      wrapper.setState({ chartType: mockChartTypeTB });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
```
could we be more specific and check that the hints array contains the correct hint...

ascott replied:
let's check this hint string as well like you did here: https://github.com/airbnb/superset/pull/2811/files#diff-5fd795ed9490b587cec77d05effde1d1R169

graceguo-supercat replied:
fixed

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
+      columnsStub.returns(mockColumns);
+      wrapper.setState({ chartType: null });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
```
could we be more specific and check that the hints array contains the correct hint...

graceguo-supercat replied:
now I wrap validation hint message text into constants file, so that in unit test I can check hint message matched with specific content.

ascott replied:
👍

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
+
+      wrapper.instance().visualize();
+      expect(window.open.callCount).to.equal(1);
+      datasourceSpy.restore();
```
this line could be added to the `afterEach` block

graceguo-supercat replied:
fixed.

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -66,47 +77,230 @@ describe('VisualizeModal', () => {
+      wrapper.instance().visualize();
+      expect(window.open.callCount).to.equal(0);
+      expect(notify.error.callCount).to.equal(1);
+      datasourceSpy.restore();
```
this line could be added to the `afterEach` block

ascott commented:
a few comments, but generally looks good!

Commit:
fix per code review comments

Commit:
fix per code review comments

ascott commented on superset/assets/javascripts/SqlLab/constants.js:
```diff
@@ -22,3 +22,10 @@ export const TIME_OPTIONS = [
   '1 year ago',
 ];
+
+export const VALIDATION_ERRORS = {
```
this might more clear as `VISUALIZE_VALIDATION_ERRORS`

ascott commented on superset/assets/spec/javascripts/sqllab/VisualizeModal_spec.jsx:
```diff
@@ -66,47 +78,232 @@ describe('VisualizeModal', () => {
+        '&': 1,
+      });
+      wrapper.instance().validate();
+      expect(wrapper.state().hints).to.have.length(1);
```
can we assert that the correct hint is in the array like you did here: https://github.com/airbnb/superset/pull/2811/files#diff-5fd795ed9490b587cec77d05effde1d1R169

graceguo-supercat replied:
fixed

ascott commented:
LGTM 🚢

Commit:
Merge branch 'master' into gg-VisualModalTests
